### PR TITLE
fix bottom_sheet closing with upwards motion

### LINKED ISSUE #138839: showModalBottomSheet closes when dragging upwards to expand
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [x] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Steps to reproduce

1. Open modal sheet
2. Drag sheet down near the bottom
3. Drag up fast and release the finger inside the area occupied by the bottom half of the extended sheet.

### Expected results

The sheet should expand after dragging it up.

### Actual results

The sheet closes.

### Code sample

<details><summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MainApp());
}

class MainApp extends StatelessWidget {
  const MainApp({super.key});

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      home: HomePage(),
    );
  }
}

class HomePage extends StatelessWidget {
  const HomePage({super.key});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: ElevatedButton(
          child: const Text('Open sheet'),
          onPressed: () => _showModalSheet(context),
        ),
      ),
    );
  }

  Future<dynamic> _showModalSheet(BuildContext context) {
    return showModalBottomSheet(
      isScrollControlled: true,
      context: context,
      builder: (context) {
        return SizedBox(
          height: MediaQuery.sizeOf(context).height * 0.8,
          child: const Padding(
            padding: EdgeInsets.all(24),
            child: Stack(
              children: [
                Text(
                    'Drag the sheet down and then drag it up fast. Releasing the sheet near the bottom, under half of the sheet height, should expand the sheet to full height. Instead, it closes the sheet.'),
                Center(
                  child: Text('Release finger under this | 0.5*height '),
                ),
              ],
            ),
          ),
        );
      },
    );
  }
}

```

</details>

### Screenshots or Video

<summary>Screenshots / Video demonstration</summary>

https://github.com/flutter/flutter/assets/114481466/16ca2071-f287-4c43-a437-3a3ea28fadd6

### Logs

_No response_

### Flutter Doctor output

<details><summary>Doctor output</summary>

```console
[✓] Flutter (Channel stable, 3.16.0, on macOS darwin-arm64)
    • Flutter version 3.16.0 on channel stable at /Users/_/Developer/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision db7ef5bf9f (6 days ago), 2023-11-15 11:25:44 -0800
    • Engine revision 74d16627b9
    • Dart version 3.2.0
    • DevTools version 2.28.2

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.1)
    • Android SDK at /Users/_/Library/Android/sdk
    • Platform android-33, build-tools 33.0.1
    • ANDROID_HOME = /Users/_/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14B47b
    • CocoaPods version 1.13.0

[✓] Android Studio (version 2022.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)

[✓] VS Code (version 1.84.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.76.0

[✓] Connected device

[✓] Network resources
    • All expected network resources are available.
```

</details>

### PR DESCRIPTION
**Pull request by miguelmato**
Fix for bottom sheet closing when being dragged up near the bottom. 
Fixes #138839 

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

### CODE REVIEW

File: packages/flutter/lib/src/material/bottom_sheet.dart [hunk 1]
```diff
@@ -296,7 +296,7 @@ class _BottomSheetState extends State<BottomSheet> {
       dragHandleMaterialState.remove(MaterialState.dragged);
     });
     bool isClosing = false;
-    if (details.velocity.pixelsPerSecond.dy > _minFlingVelocity) {
+    if (details.velocity.pixelsPerSecond.dy.abs() > _minFlingVelocity) {
```
Comment (miguelmato) [inline packages/flutter/lib/src/material/bottom_sheet.dart:299]: Only swiping down (positive y) velocity was considered even though the following code was expecting both directions as valid

**Comment by miguelmato** [comment]
cc @HansMuller the fix is very simple, but should I add a test regardless?

**Comment by HansMuller** [comment]
Yes, a regression test is needed; we don't want this behavior to come back.

File: packages/flutter/test/material/bottom_sheet_test.dart [hunk 2]
```diff
@@ -2263,6 +2263,99 @@ void main() {
           tester.widget(find.byType(ModalBarrier).last);
       expect(modalBarrier.semanticsLabel, MaterialLocalizations.of(scaffoldKey.currentContext!).scrimLabel);
     });
+
+    testWidgetsWithLeakTracking('Swiping up an almost contracted sheet should expand it', (WidgetTester tester) async {
+
+      Future<void> flingFrom(
```
Comment (miguelmato) [inline packages/flutter/test/material/bottom_sheet_test.dart:2269]: this function is copied from source. the initial offset is useful to simulate a drag before the fling, but a bug in the implementation makes the initialoffset not functional. It needs to do at least 2 move events: first to detect a drag, second to actually drag. 

Smelly code duplication, need advice on how to handle this

**Comment by HansMuller** [comment]
@dkwingsmt - could you take a look at this? The fix in the framework is quite small but there's a rather large change in packages/flutter_test/lib/src/controller.dart

File: packages/flutter_test/lib/src/controller.dart [hunk 3]
```diff
@@ -1238,11 +1238,92 @@ abstract class WidgetController {
       double timeStamp = 0.0;
       double lastTimeStamp = timeStamp;
       await sendEventToBinding(testPointer.down(startLocation, timeStamp: Duration(microseconds: timeStamp.round())));
+
       if (initialOffset.distance > 0.0) {
-        await sendEventToBinding(testPointer.move(startLocation + initialOffset, timeStamp: Duration(microseconds: timeStamp.round())));
+        final double xSign = initialOffset.dx.sign;
```
Comment (dkwingsmt) [inline packages/flutter_test/lib/src/controller.dart:1243]: I tried removing all changes here and your new test still passes. Is it still necessary?

File: packages/flutter/test/material/bottom_sheet_test.dart [hunk 4]
```diff
@@ -2288,6 +2288,49 @@ void main() {
...
+    await tester.flingFrom(
+      tester.getCenter(find.text('BottomSheet')),
+      const Offset(0, -100),
+      800, // _kminFlingVelocity == 700 pixels/second
+      initialOffset: const Offset(0, sheetHeight * 0.8),
+      initialOffsetDelay: const Duration(seconds: 1),
+    );
+
+    await tester.pumpAndSettle();
+    expect(showBottomSheetThenCalled, isFalse);
```
Comment (dkwingsmt) [inline packages/flutter/test/material/bottom_sheet_test.dart:2333]:
```suggestion
    expect(showBottomSheetThenCalled, isFalse);
    });
```

Also indentation is wrong.

File: packages/flutter_test/lib/src/controller.dart (hunk 3 above)
Comment (miguelmato) [reply packages/flutter_test/lib/src/controller.dart:1243]: Yes, basically the flingFrom method has an initialOffset paremeter very useful when simulating a swipe/fling after a desplacement, but the current implementation makes initialOffset non funtional, so no displacement is actually performed. [Current behavior of flingFrom.initialOffset ](https://github.com/flutter/flutter/issues/139455#issuecomment-1838323397)

That's why the test passes with the old implementation, because the sheet does not move to the bottom to trigger the edge condition that causes it to close when it should expand again to full height. 

The changes in the controller is to copy the code from dragFrom, that in order to drag a pointer, it has to be divided in smaller distances.

Comment (dkwingsmt) [reply packages/flutter_test/lib/src/controller.dart:1243]: Thanks for finding this problem. Can you write a test in `controller_test.dart`? This kind of utility code should also be guarded by tests.

File: packages/flutter/test/material/bottom_sheet_test.dart (hunk 4 above)
Comment (miguelmato) [reply packages/flutter/test/material/bottom_sheet_test.dart:2333]: Oops, my bad. I also notices some extra white lines, will fix the syntax later today

File: packages/flutter_test/lib/src/controller.dart (hunk 3 above)
Comment (dkwingsmt) [reply packages/flutter_test/lib/src/controller.dart:1243]: Yep. Just precisely test what you found wrong.

**Comment by dkwingsmt** [comment]
Hi, any updates on the PR?

**Comment by Piinks** [comment]
Hi @miguelmato, thank you for your contribution. Since we have not heard back in a while, I am going to close this PR to remove it from the review queue. If you would like to continue working on this, please feel free to re-open it when you are ready. Thanks!
